File: front/src/api/datasets.ts
import { AxiosInstance } from "axios";
import { z } from "zod";

import { DatasetSchema, RecordingStateSchema } from "@/schemas";

import { GetManySchema, Page } from "./common";

import type { Dataset, RecordingState } from "@/types";

export const DatasetFilterSchema = z.object({
  search: z.string().optional(),
});

export type DatasetFilter = z.input<typeof DatasetFilterSchema>;

export const DatasetCreateSchema = z.object({
  uuid: z.string().uuid().optional(),
  name: z.string().min(1),
  audio_dir: z.string(),
  description: z.string().optional(),
});

export type DatasetCreate = z.input<typeof DatasetCreateSchema>;

export const DatasetUpdateSchema = z.object({
  name: z.string().min(1).optional(),
  description: z.string().optional(),
});

export type DatasetUpdate = z.input<typeof DatasetUpdateSchema>;

export const DatasetPageSchema = Page(DatasetSchema);

export type DatasetPage = z.infer<typeof DatasetPageSchema>;

export const GetDatasetsQuerySchema = z.intersection(
  GetManySchema,
  DatasetFilterSchema,
);

export type GetDatasetsQuery = z.input<typeof GetDatasetsQuerySchema>;

const DEFAULT_ENDPOINTS = {
  getMany: "/api/v1/datasets/",
  create: "/api/v1/datasets/",
  state: "/api/v1/datasets/detail/state/",
  get: "/api/v1/datasets/detail/",
  update: "/api/v1/datasets/detail/",
  delete: "/api/v1/datasets/detail/",
  downloadJson: "/api/v1/datasets/detail/download/json/",
  downloadCsv: "/api/v1/datasets/detail/download/csv/",
  import: "/api/v1/datasets/import/",
};

export function registerDatasetAPI({
  instance,
  endpoints = DEFAULT_ENDPOINTS,
  baseUrl = "",
}: {
  instance: AxiosInstance;
  endpoints?: typeof DEFAULT_ENDPOINTS;
  baseUrl?: string;
}) {
  async function getMany(query: GetDatasetsQuery): Promise<DatasetPage> {
    const params = GetDatasetsQuerySchema.parse(query);
    const { data } = await instance.get(endpoints.getMany, { params });
    return DatasetPageSchema.parse(data);
  }

  async function create(data: DatasetCreate): Promise<Dataset> {
    const body = DatasetCreateSchema.parse(data);
    const { data: res } = await instance.post(endpoints.create, body);
    return DatasetSchema.parse(res);
  }

  async function get(uuid: string): Promise<Dataset> {
    const { data } = await instance.get(endpoints.get, {
      params: { dataset_uuid: uuid },
    });
    return DatasetSchema.parse(data);
  }

  async function getDatasetState(uuid: string): Promise<RecordingState[]> {
    const { data } = await instance.get(endpoints.state, {
      params: { dataset_uuid: uuid },
    });
    return z.array(RecordingStateSchema).parse(data);
  }

  async function updateDataset(
    dataset: Dataset,
    data: DatasetUpdate,
  ): Promise<Dataset> {
    const body = DatasetUpdateSchema.parse(data);
    const { data: res } = await instance.patch(endpoints.update, body, {
      params: { dataset_uuid: dataset.uuid },
    });
    return DatasetSchema.parse(res);
  }

  async function deleteDataset(dataset: Dataset): Promise<Dataset> {
    const { data } = await instance.delete(endpoints.delete, {
      params: { dataset_uuid: dataset.uuid },
    });
    return DatasetSchema.parse(data);
  }

  function getDownloadUrl(dataset: Dataset, format: "json" | "csv"): string {
    if (format === "json") {
      return `${baseUrl}${endpoints.downloadJson}?dataset_uuid=${dataset.uuid}`;
    } else {
      return `${baseUrl}${endpoints.downloadCsv}?dataset_uuid=${dataset.uuid}`;
    }
  }

  async function importDataset(data: FormData): Promise<Dataset> {
    const { data: res } = await instance.post(endpoints.import, data);
    return DatasetSchema.parse(res);
  }

  return {
    getMany,
    create,
    get,
    getState: getDatasetState,
    update: updateDataset,
    delete: deleteDataset,
    getDownloadUrl,
    import: importDataset,
  } as const;
}

File: back/src/whombat/routes/datasets.py
"""REST API routes for datasets."""

import datetime
import logging
from io import StringIO
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Body, Depends, UploadFile
from fastapi.responses import Response, StreamingResponse
from pydantic import DirectoryPath
from soundevent.io.aoef import DatasetObject
from sqlalchemy.exc import IntegrityError

from whombat import api, exceptions, schemas
from whombat.filters.datasets import DatasetFilter
from whombat.routes.dependencies import Session, WhombatSettings
from whombat.routes.types import Limit, Offset

__all__ = [
    "dataset_router",
]

dataset_router = APIRouter()

logger = logging.getLogger(__name__)


@dataset_router.get(
    "/detail/",
    response_model=schemas.Dataset,
)
async def get_dataset(
    session: Session,
    dataset_uuid: UUID,
):
    """Get a dataset by UUID."""
    return await api.datasets.get(session, dataset_uuid)


@dataset_router.get(
    "/",
    response_model=schemas.Page[schemas.Dataset],
)
async def get_datasets(
    session: Session,
    filter: Annotated[
        DatasetFilter,  # type: ignore
        Depends(DatasetFilter),
    ],
    limit: Limit = 10,
    offset: Offset = 0,
):
    """Get a page of datasets."""
    datasets, total = await api.datasets.get_many(
        session,
        limit=limit,
        offset=offset,
        filters=[filter],
    )

    return schemas.Page(
        items=datasets,
        total=total,
        offset=offset,
        limit=limit,
    )


@dataset_router.post(
    "/",
    response_model=schemas.Dataset,
)
async def create_dataset(
    session: Session,
    dataset: schemas.DatasetCreate,
):
    """Create a new dataset."""
    created = await api.datasets.create(
        session,
        name=dataset.name,
        description=dataset.description,
        dataset_dir=dataset.audio_dir,
    )
    await session.commit()
    return created


@dataset_router.patch(
    "/detail/",
    response_model=schemas.Dataset,
)
async def update_dataset(
    session: Session,
    dataset_uuid: UUID,
    data: schemas.DatasetUpdate,
):
    """Update a dataset."""
    dataset = await api.datasets.get(session, dataset_uuid)
    updated = await api.datasets.update(session, dataset, data)
    await session.commit()
    return updated


@dataset_router.get(
    "/detail/state/",
    response_model=list[schemas.DatasetFile],
)
async def get_file_state(
    session: Session,
    dataset_uuid: UUID,
):
    """Get the status of the files in a dataset."""
    dataset = await api.datasets.get(session, dataset_uuid)
    return await api.datasets.get_state(session, dataset)


@dataset_router.delete(
    "/detail/",
    response_model=schemas.Dataset,
)
async def delete_dataset(
    session: Session,
    dataset_uuid: UUID,
):
    """Delete a dataset."""
    dataset = await api.datasets.get(session, dataset_uuid)

    try:
        deleted = await api.datasets.delete(session, dataset)
    except IntegrityError as error:
        raise exceptions.DataIntegrityError(
            "Cannot delete this dataset because it is currently in use. "
            "This dataset may be associated with active annotation projects "
            "or other processes. Please ensure that the dataset is not being "
            "used in any active tasks before attempting to delete it."
        ) from error


    await session.commit()
    return deleted


@dataset_router.get(
    "/detail/download/json/",
    response_model=DatasetObject,
)
async def download_dataset_json(
    session: Session,
    dataset_uuid: UUID,
    settings: WhombatSettings,
):
    """Export a dataset."""
    whombat_dataset = await api.datasets.get(session, dataset_uuid)
    obj = await api.datasets.export_dataset(
        session,
        whombat_dataset,
        audio_dir=settings.audio_dir,
    )
    filename = f"{whombat_dataset.name}_{obj.created_on.isoformat()}.json"
    return Response(
        obj.model_dump_json(),
        media_type="application/json",
        status_code=200,
        headers={"Content-Disposition": f"attachment; filename={filename}"},
    )


@dataset_router.get(
    "/detail/download/csv/",
)
async def download_dataset_csv(
    session: Session,
    dataset_uuid: UUID,
):
    """Export the dataset recordings in csv format."""
    dataset = await api.datasets.get(session, dataset_uuid)
    df = await api.datasets.to_dataframe(session, dataset)
    buffer = StringIO()
    df.to_csv(buffer, index=False)
    buffer.seek(0)
    filename = f"{dataset.name}_{datetime.datetime.now()}.csv"
    return StreamingResponse(
        buffer,
        media_type="text/csv",
        status_code=200,
        headers={"Content-Disposition": f"attachment; filename={filename}"},
    )


@dataset_router.post(
    "/import/",
    response_model=schemas.Dataset,
)
async def import_dataset(
    settings: WhombatSettings,
    session: Session,
    dataset: UploadFile,
    audio_dir: Annotated[DirectoryPath, Body()],
):
    """Import a dataset."""
    if not audio_dir.exists():
        raise FileNotFoundError(f"Audio directory {audio_dir} does not exist.")

    return await api.datasets.import_dataset(
        session,
        dataset.file,
        dataset_audio_dir=audio_dir,
        audio_dir=settings.audio_dir,
    )
